mypy/typeanal.py

```diff
@@ -312,11 +307,17 @@ def visit_instance(self, t: Instance) -> Type:
     def visit_type_var(self, t: TypeVarType) -> Type:
         return t
 
-    def visit_callable_type(self, t: CallableType) -> Type:
-        return t.copy_modified(arg_types=self.anal_array(t.arg_types, nested=False),
-                               ret_type=self.anal_type(t.ret_type, nested=False),
-                               fallback=t.fallback or self.builtin_type('builtins.function'),
-                               variables=self.anal_var_defs(t.variables))
+    def visit_callable_type(self, t: CallableType, nested: bool = True) -> Type:
+        with self.tvar_scope_frame():
```

Comment:
The above line is the punchline of this diff.  Make a layer of type variable scope when analyzing a callable type, so it can bind its own type variables if it needs to.